funcx_endpoint/funcx_endpoint/endpoint/rabbit_mq/task_queue_subscriber.py

```diff
@@ -0,0 +1,401 @@
+
+    def reconnect(self):
+        """Will be invoked by the IOLoop timer if the connection is
+        closed. See the on_connection_closed method.
+
+        """
+        # This is the old connection IOLoop instance, stop its ioloop
+        self._connection.ioloop.stop()
+
+        if not self._closing:
+
+            # Create a new connection
+            self._connection = self.connect()
+
+            # There is now a new connection, needs a new ioloop to run
+            self._connection.ioloop.start()
+
+    def open_channel(self):
+        """Open a new channel with RabbitMQ by issuing the Channel.Open RPC
+        command. When RabbitMQ responds that the channel is open, the
+        on_channel_open callback will be invoked by pika.
+
+        """
+        logger.info("Creating a new channel")
+        self._connection.channel(on_open_callback=self._on_channel_open)
+
+    def _on_channel_open(self, channel):
+        """This method is invoked by pika when the channel has been opened.
+        The channel object is passed in so we can make use of it.
+
+        Since the channel is now open, we'll declare the exchange to use.
+
+        :param pika.channel.Channel channel: The channel object
+
+        """
+        logger.info(f"Channel opened: {channel}")
+        self._channel = channel
+        self.add_on_channel_close_callback()
+        self.setup_exchange(self.exchange)
+
+    def add_on_channel_close_callback(self):
+        """This method tells pika to call the on_channel_closed method if
+        RabbitMQ unexpectedly closes the channel.
+
+        """
+        logger.info("Adding channel close callback")
+        self._channel.add_on_close_callback(self._on_channel_closed)
+
+    def _on_channel_closed(self, channel, exception):
+        """Invoked by pika when RabbitMQ unexpectedly closes the channel.
+        Channels are usually closed if you attempt to do something that
+        violates the protocol, such as re-declare an exchange or queue with
+        different parameters. In this case, we'll close the connection
+        to shutdown the object.
+
+        :param pika.channel.Channel: The closed channel
+        :param Exception: exception from pika
+        """
+        logger.warning(f"Channel:{channel} was closed: ({exception}")
+        try:
+            logger.warning("Raising exception:")
+            raise exception
+        except pika.exceptions.ChannelClosedByBroker:
+            if "exclusive use" in exception.reply_text:
+                logger.exception(
+                    "Channel closed by RabbitMQ broker due to exclusive "
+                    "ownership by an active endpoint"
+                )
+                logger.warning("Channel will close without connection retry")
+                self._closing = True
+        except Exception as e:
+            logger.exception(
+                f"Channel closed with code:{e.reply_code}, error:{e.reply_text}"
+            )
```

Comment:
Intentionally raising and squashing an exception is a very odd way to structure this, and obscures the fact that no exception is ever intentionally raised from this method (because you're catching and discarding it). Just use an `if isinstance` check and say what you mean to say.

I'm also a bit suspicious of this not handling `ChannelClosedByClient`, which is documented as being received here if the channel is closed by client-side code. I think that should at least be handled and explicitly ignored.

A suggested update to this code below:

```suggestion
        if isinstance(exception, pika.exceptions.ChannelClosedByBroker):
            if "exclusive use" in exception.reply_text:
                logger.exception(
                    "Channel closed by RabbitMQ broker due to exclusive "
                    "ownership by an active endpoint"
                )
                logger.warning("Channel will close without connection retry")
                self._closing = True
        if isinstance(exception, pika.exceptions.ChannelClosedByClient):
            logger.debug("detected channel closed by client")
        else:
            logger.error(
                f"Channel closed with code:{e.reply_code}, error:{e.reply_text}"
            )
```